BraveRewardsUI/Extensions/BraveLedgerExtensions.swift

```diff
@@ -117,6 +120,90 @@ extension BraveLedger {
       })
     }
   }
+  
+  private var paymentId: String? {
+    var id: String?
+    rewardsInternalInfo { info in
+      id = info?.paymentId
+    }
+    return id
+  }
+  
+  public func setupDeviceCheckEnrollment(_ client: DeviceCheckClient, completion: @escaping () -> Void) {
+    // Enroll in DeviceCheck
+    client.generateToken { [weak self] (token, error) in
+      guard let self = self else { return }
+      if let error = error {
+        log.error("Failed to generate DeviceCheck token: \(error)")
+        completion()
+        return
+      }
+      let paymentId = self.paymentId ?? ""
+      client.generateEnrollment(paymentId: paymentId, token: token) { registration, error in
+        if let error = error {
+          log.error("Failed to enroll in DeviceCheck: \(error)")
+          completion()
+          return
+        }
+        guard let registration = registration else { return }
+        client.registerDevice(enrollment: registration) { error in
+          if let error = error {
+            log.error("Failed to register device with mobile attestation server: \(error)")
+            completion()
+            return
+          }
+        }
+      }
+    }
+  }
+  
+  public func claimPromotion(_ promotion: Promotion, completion: @escaping (_ success: Bool) -> Void) {
+    guard let paymentId = self.paymentId else { return }
+    let deviceCheck = DeviceCheckClient(environment: BraveLedger.environment)
+    let group = DispatchGroup()
+    if !DeviceCheckClient.isDeviceEnrolled() {
+      group.enter()
+      setupDeviceCheckEnrollment(deviceCheck) {
+        if !DeviceCheckClient.isDeviceEnrolled() {
+          completion(false)
+          return
+        }
+        group.leave()
+      }
+    }
+    group.notify(queue: .main) {
+      deviceCheck.generateAttestation(paymentId: paymentId) { (attestation, error) in
+        guard let attestation = attestation else {
+          completion(false)
+          return
+        }
+        self.claimPromotion(attestation.publicKeyHash) { result, noonce in
+          if result != .ledgerOk {
+            completion(false)
+            return
+          }
+          
+          deviceCheck.generateAttestationVerification(nonce: noonce) { verification, error in
+            guard let verification = verification else {
+              completion(false)
+              return
+            }
+            
+            let solution = PromotionSolution()
+            solution.noonce = noonce
+            solution.signature = verification.signature
+            do {
+              solution.blob = try verification.attestationBlob.bsonData().base64EncodedString()
+            } catch { }
```

Comment:
scary to have empty catch all, can we add a log at least? Or is it too spammy here

Reply:
No I can totally add a log at the very least, thanks for catching.